### DISCUSSION TITLE
What is the best way to extend QTree?

### DISCUSSION THREAD

**Discussion by flowt-au** [quasarframework/quasar#12085] [General - Components / Directives / etc]
I am upgrading my Q1 app to Q2. In Q1 I added some extra tree methods to "extend" the qtree using a dubious technique, that worked.
eg cascade and getParentNode functions (trimmed for brevity):
```
this.$refs['mytree'].cascade = function (startingNodeArray, callback) {}
this.$refs['mytree].getParentNode = function (nodeid) {}
```
`getParentNode ()` accessed `this.meta` as did a number of my other custom functions. In Q2, `meta` is private. 

So, I know I need to learn more about Vue 3 and Q2 which I am more than happy to do. I see there are a few posts suggesting different techniques and I tried a few things but I could not work out how I would go about accessing the meta computed ( `const meta = computed(() => {})`) as per `ui/src/components/tree/QTree.js`

My question is: what is the overall approach pattern for extending Quasar components in a way that I can access the private data and methods? I need some clues about what I need to study and learn so I can reimplement my custom functions in Q2.

I could obviously copy your code and make my own tree eg as an app extension or whatever, but I was hoping for a more sustainable solution that just extends whatever the current version of QTree is.

Thanks!
Murray

**Comment by metalsadman** [comment]
you can't access it outside `setup` if it was not exposed explicitly. I think making a wrapper should work, and just use a ref using the wrapped qtree methods in your added functionality of your wrapper.

**Reply by flowt-au** [reply]
Thanks @metalsadman. As you say I cant access `meta` because it isn't exposed. Similarly, the other functions that the meta computed function uses are also not exposed. So, I can't really "extend" QTree because my added methods need `meta` and its dependents. 

>I think making a wrapper should work

Sorry, I am not with you. How would I wrap `meta` and the other related functions since they are internal to QTree?

Thanks for your help, sorry if I am slow to catch on here.

Cheers,
Murray

**Comment by flowt-au** [comment]
I just made a feature request and explanation: #12125 
:-)
